drivers/tty/serdev/pibridge.c

```diff
@@ -0,0 +1,409 @@
+
+	/* RTS is used to drive Transmit Enable pin, hence no flow control */
+	serdev_device_set_flow_control(serdev, false);
+
+	return serdev_device_set_parity(serdev, SERDEV_PARITY_EVEN);
+}
+
+
+static int pibridge_probe(struct serdev_device *serdev)
+{
+	struct device *dev = &serdev->dev;
+	struct pibridge *pi;
+	int ret;
+
+	pi = devm_kzalloc(dev, sizeof(*pi), GFP_KERNEL);
+	if (!pi)
+		return -ENOMEM;
+
+	pibridge_s = pi;
+	pi->serdev = serdev;
+
+	serdev_device_set_drvdata(serdev, pi);
+	serdev_device_set_client_ops(serdev, &pibridge_serdev_ops);
+
+	mutex_init(&pi->lock);
+	init_waitqueue_head(&pi->read_queue);
+
+	ret = kfifo_alloc(&pi->read_fifo, REV_PI_RECV_BUFFER_SIZE, GFP_KERNEL);
+	if (ret)
+		return ret;
+
+	ret = serdev_device_open(serdev);
+	if (ret)
+		goto err_kfifo_free;
+
+	ret = pibridge_parse_dt(serdev);
+	if (ret)
+		goto err_serdev_close;
+
+	return 0;
+
+err_serdev_close:
+	serdev_device_close(serdev);
+err_kfifo_free:
+	kfifo_free(&pi->read_fifo);
+	return ret;
+}
+
+static void pibridge_remove(struct serdev_device *serdev)
+{
+	struct pibridge *pi = serdev_device_get_drvdata(serdev);
+
+	serdev_device_close(serdev);
+
+	kfifo_free(&pi->read_fifo);
+};
+
+#ifdef CONFIG_OF
+static const struct of_device_id pibridge_of_match[] = {
+	{ .compatible = "kunbus,pibridge" },
+	{},
+};
+MODULE_DEVICE_TABLE(of, pibridge_of_match);
+#endif
+
+static struct serdev_device_driver pibridge_driver = {
+	.driver	= {
+		.name		= "pibridge",
+		.of_match_table	= of_match_ptr(pibridge_of_match),
+	},
+	.probe	= pibridge_probe,
+	.remove	= pibridge_remove,
+};
+module_serdev_device_driver(pibridge_driver);
+
+
+/*****************/
+
+int pibridge_send(u8 *buf, u16 len)
+{
+	struct pibridge *pi = pibridge_s;
+	struct serdev_device *serdev = pi->serdev;
+	int ret;
+
+	ret = serdev_device_write(serdev, buf, len, MAX_SCHEDULE_TIMEOUT);
+	if (ret < 0)
+		pb_err(25, "write to serdev error(len:%d)\n", len);
+
+	/* read fifo may contain stale bytes */
+	mutex_lock(&pi->lock);
+	kfifo_reset(&pi->read_fifo);
+	mutex_unlock(&pi->lock);
+
+	serdev_device_wait_until_sent(serdev, 0);
+
+	return ret;
+}
+
+int pibridge_recv_timeout(u8 *buf, u16 len, u16 timeout)
+{
+	struct pibridge *pi = pibridge_s;
+	int jiffies;
+	int ret;
+
+	jiffies = wait_event_timeout(pi->read_queue, kfifo_len(&pi->read_fifo) >= len,
+			msecs_to_jiffies(timeout));
+	mutex_lock(&pi->lock);
+	if (jiffies){
+		ret = kfifo_out(&pi->read_fifo, buf, len);
+	} else {
+		ret = 0;
+	}
+	mutex_unlock(&pi->lock);
+	if (ret < len)
+		pb_err(25, "receive message error(len:%d, ret:%d, jiffies:%d, fifo:%d)\n",
+				len, ret, jiffies, kfifo_len(&pi->read_fifo));
+	return ret;
+}
+
+int pibridge_recv(u8 *buf, u16 len)
+{
+	/* using default timeout REV_PI_IO_TIMEOUT */
+	return pibridge_recv_timeout(buf, len, REV_PI_IO_TIMEOUT);
+}
+
+#pragma pack(1)
+struct pibridge_pkthdr_gate{
+	u8	dst;
+	u8	src;
+	u16	cmd;
+	u16	seq;
+	u8	len;
+};
+
+u8 pibridge_crc8(u8 base, u8 *data, u16 len)
```

Comment:
Can be declared `static`. Probably applies to other functions as well.

Reply:
Agreed